### DISCUSSION TITLE
Callback-`Protocol` for the Union of all `Callable`-types that only take keyword arguments.

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1468] [Q&A]
Is it possible to express the type of the union of all `Callable`-types that only take keyword arguments (*)? I tried the following, but it gives both false positives and false negatives:

```python
from typing import Protocol, ParamSpec

P = ParamSpec("P")

class KWOnlyFunc(Protocol[P]):
    """Protocol for callables that accept only keyword-arguments."""
    def __call__(self, **kwargs: P.kwargs) -> None: ...

def false0(x: int) -> None: ...
def false1(x: int, **kwargs: int) -> None: ...
def false2(*args: int, **kwargs: int) -> None: ...
def false3(*args: int, z: int, **kwargs: int) -> None: ...
def false4(*args: int, z: int=0, **kwargs: int) -> None: ...
def true0(**kwargs: int) -> None: ...
def true1(*, z: int) -> None: ...
def true2(*, z: int, **kwargs: int) -> None: ...
def true3(*, z: int=0) -> None: ...
def true4(*, z: int=0, **kwargs: int) -> None: ...

f0: KWOnlyFunc = false0 # ✔  true negative
f1: KWOnlyFunc = false1 # ✔  true negative
f2: KWOnlyFunc = false2 # ✘  false postive
f3: KWOnlyFunc = false3 # ✔  true negative
f4: KWOnlyFunc = false4 # ✘  false postive

t0: KWOnlyFunc = true0  # ✔  true positive
t1: KWOnlyFunc = true1  # ✘  false negative
t2: KWOnlyFunc = true2  # ✘  false negative
t3: KWOnlyFunc = true3  # ✘  false negative
t4: KWOnlyFunc = true4  # ✔  true positive
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=6307ab826f8a3bf3ee6274999f15dde6

(*) I.e. They satisfy the following `TypeGuard`

```python
def is_kw_only_func(f: Callable) -> TypeGuard[KWOnlyFunc]:
    sig = inspect.signature(f)
    return all(p.kind in (p.KW_ONLY, p.VAR_KEYWORD) for p in sig.parameters.values())
```

**Comment by erictraut** [comment]
It's not legal to specify a `**kwargs: P.kwargs` parameter without a corresponding `*args: P.args`. PEP 612 is [very clear on this point](https://peps.python.org/pep-0612/#id1). Mypy should generate an error for your invalid protocol definition, but it doesn't.

**Reply by randolf-scholz** [reply]
Do you know if there are any plans for adding a map-type variadic (as a counterpart to TypeVarTuple?), this would solve my problem.

**Reply by randolf-scholz** [reply]
It's also noteworthy that the rationale given in the PEP for disallowing `P.kwargs` without `P.args` is not really applicable here. Seems like this option was overlooked.

**Reply by Daverball** [reply]
PEP612 specifically mentions `bound` as not having been specified/implemented yet for `ParamSpec`, even though for all intents and purposes it is a `TypeVar`, so it should be added in the future.

So in a future PEP `bound` on `ParamSpec` could be used to provide an upper bound on the shape of the `ParamSpec` which would allow you to restrict the shape to contain no positional parameters. But I think what needs to happen before that is to introduce a notation for all the different kinds of arguments in a `ParamSpec`, since right now you can only bind unnamed parameters and can't specify optional parameters or variadic arguments either.

Something like PEP695 but for function signatures, rather than generics.
